pkg/cgroup/cgroup.go

```diff
@@ -14,73 +12,352 @@ import (
+	)
 	if err != nil {
-		return -1, fmt.Errorf("could not open cgroups file %s: %w", cgroupsFile, err)
+		return err
+	}
+
+	// 2. discover where cgroup is mounted
+	c.mountpoint = c.mounted.GetMountpoint()
+
+	return nil
+}
+
+func (c *CgroupV2) destroy() error {
+	return c.mounted.Umount()
+}
+
+func (c *CgroupV2) GetMountPoint() string {
+	return c.mountpoint
+}
+
+func (c *CgroupV2) getDefaultHierarchyID() int {
+	return c.hid
+}
+
+//
+// General
+//
+
+func GetCgroupDefaultVersion() (CgroupVersion, error) {
+	// 1st Method: already mounted cgroupv1 filesystem
+
+	ok, err := IsCgroupV1MountedAndDefault()
+	if err != nil {
+		return -1, err
+	}
+	if ok {
+		return CgroupVersion1, nil
+	}
+
+	//
+	// 2nd Method: From cgroup man page:
+	// ...
+	// 2. The unique ID of the cgroup hierarchy on which this
+	//    controller is mounted. If multiple cgroups v1
+	//    controllers are bound to the same hierarchy, then each
+	//    will show the same hierarchy ID in this field.  The
+	//    value in this field will be 0 if:
+	//
+	//    a) the controller is not mounted on a cgroups v1
+	//       hierarchy;
+	//    b) the controller is bound to the cgroups v2 single
+	//       unified hierarchy; or
+	//    c) the controller is disabled (see below).
+	// ...
+
+	var value int
+
+	file, err := os.Open(procCgroups)
+	if err != nil {
+		return -1, CouldNotOpenFile(procCgroups, err)
 	}
 	defer file.Close()
 
 	scanner := bufio.NewScanner(file)
-	for i := 1; scanner.Scan(); i++ {
-		sline := strings.Fields(scanner.Text())
-		if len(sline) < 2 || sline[0] != CgroupV1Controller {
+	for scanner.Scan() {
+		line := strings.Fields(scanner.Text())
+		if line[0] != CgroupDefaultController {
 			continue
 		}
-		intID, err := strconv.Atoi(sline[1])
-		if err != nil || intID < 0 {
-			return -1, fmt.Errorf("error parsing %s: %w", cgroupsFile, err)
+		value, err = strconv.Atoi(line[1])
+		if err != nil || value < 0 {
+			return -1, ErrorParsingFile(procCgroups, err)
 		}
-		return intID, nil
 	}
-	return -1, fmt.Errorf("couldn't determine cgroup v1 hierarchy id")
+
+	if value == 0 { // == (a), (b) or (c)
+		return CgroupVersion2, nil
+	}
+
+	return CgroupVersion1, nil
 }
 
-// MountCgroupV1Cpuset mounts cpuset cgroup controller to the container's
-// current working directory this is done in order to detect host cgroup data
-// when running tracee from a container in cgroupsv1 since the function uses the
-// mount syscall, it requires CAP_SYS_ADMIN
-func MountCgroupV1Cpuset(mountPath string) error {
-	mp, _ := mount.IsMountpoint(mountPath, CgroupV1FsType)
+// IsCgroupV1MountedAndDefault tests if cgroupv1 is mounted and is the default
+// cgroup version being used by the running environment. It does so by checking
+// the existance of a "release_agent" file in the cpuset cgroup controller.
+//
+// https://www.kernel.org/doc/Documentation/cgroup-v1/cgroups.txt
+func IsCgroupV1MountedAndDefault() (bool, error) {
+	_, err := os.Stat(CgroupReleaseAgent)
```

Comment:
the method we used was to check for the existance of cgroup.controller (this is also used by runc) - shouldn't we continue with that?

Reply:
Actually I almost copied the function we had and it was:
```diff
-func RunsOnContainerV1() (bool, error) {
-       _, err := os.Stat(CgroupReleaseAgent)
-       if os.IsNotExist(err) {
-               return true, nil
-       }
-       if err != nil {
-               return false, err
-       }
-       return false, nil
-}
```

Reply:
![image](https://user-images.githubusercontent.com/7395852/203159283-4ed62c41-3efb-4723-8cfd-d4a7ead807d8.png)

Reply:
Like discussed offline, this was used to check if we run in a container or not once we already know that we are on a cgroup v1 environment (by checking for cgroup.controller existance)